src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -229,6 +229,19 @@ static term create_pair(Context *ctx, term term1, term term2)
     return ret;
 }
 
+static term reason_atom_to_error_tuple_safely(Context *ctx, term err_atom)
+{
+    if (UNLIKELY(!term_is_atom(err_atom))) {
+        AVM_LOGE(TAG, "Error term 'err_atom' was not an atom, returning 'undefined'.");
+        err_atom = UNDEFINED_ATOM;
+    }
+    if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
```

Comment:
This is probably dangerous, because it might be called from the port part of the driver — calling the RAISE_ERROR macro is really only applicable to Nifs.  You may instead need to return some marker value (in this case, something like `term_invalid_term()`), and check the return value on every call, but that would affect the complexity of your code considerably.

Alternatively, you could do the memory_ensure_free call from your port calls, or write a different function specifically for ports.  I believe the port “API” (such as it is) is designed to allow you to pass the OUT_OF_MEMORY_ATOM atom back to the caller, as a “special case”.

Reply:
Right. I should have thought about that closer. I believe the changes I made should work safely now. The port will just return `out_of_memory`, and the nif will raise `out_of_memory`. I don't think there is any safe way at that point to be able to return `{error, out_of_memory}` in the port driver.